spring-grpc-core/src/test/java/org/springframework/grpc/client/DefaultGrpcChannelFactoryShutdownTests.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2023-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.grpc.client;
+
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.grpc.client.DefaultGrpcChannelFactory.ManagedChannelWithShutdown;
+import org.springframework.test.util.ReflectionTestUtils;
+
+import io.grpc.ManagedChannel;
+
+/**
+ * Tests for shutdown process in {@link DefaultGrpcChannelFactory}.
+ */
+class DefaultGrpcChannelFactoryShutdownTests {
+
+	@Test
+	void channelsAreGracefullyShutdown() throws InterruptedException {
+		this.channelsShutdownAsExpected(false);
+	}
+
+	@Test
+	void whenChannelExceedsAwaitTimeOtherChannelsAreStillShutdownGracefully() throws InterruptedException {
+		this.channelsShutdownAsExpected(true);
+	}
+
+	private void channelsShutdownAsExpected(boolean exceedAwaitTime) throws InterruptedException {
+		var channelFactory = new DefaultGrpcChannelFactory<>(List.of(), mock());
+		channelFactory.setVirtualTargets(path -> path);
+
+		// create channels using factory and options
+		var c1 = channelFactory.createChannel("c1",
+				ChannelBuilderOptions.defaults().withShutdownGracePeriod(Duration.ofSeconds(7)));
+		var c2 = channelFactory.createChannel("c2",
+				ChannelBuilderOptions.defaults().withShutdownGracePeriod(Duration.ofSeconds(5)));
+		var c3 = channelFactory.createChannel("c3",
+				ChannelBuilderOptions.defaults().withShutdownGracePeriod(Duration.ofSeconds(10)));
+
+		// spy each channel to wait accordingly
+		var c1Spy = setupSpy(c1);
```

Comment:
I am not excited about having to get in and spy like this. However, it is a necessary evil to get some basic test coverage.

Namely, we ensure each channel gets the following:
- `shutdown` invoked
- `awaitTermination` invoked
- `shutdownNow` never invoked **EXCEPT** only on channel2 when `exceedAwaitTime = true`